bpf/bpf_lxc.c

```diff
@@ -972,19 +978,19 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx, __u32 *d
 		 */
 		ep = lookup_ip4_endpoint(ip4);
 		if (ep) {
-#ifdef ENABLE_ROUTING
 			if (ep->flags & ENDPOINT_F_HOST) {
+				if (is_defined(ENABLE_ROUTING)) {
 #ifdef HOST_IFINDEX
-				goto to_host;
-#else
-				return DROP_HOST_UNREACHABLE;
+					goto to_host;
 #endif
+					return DROP_HOST_UNREACHABLE;
+				}
+				goto pass_to_stack;
```

Comment:
I'm trying to understand this change in behavior, but I'm not sure I understand how it even used to work. AFAICS, when endpoint routes were enabled, we used to deliver packets going to the host via `ipv4_local_delivery` but that will fail as `ep->lxc_id` is 0 for the host endpoint :confused:

Reply:
> when endpoint routes were enabled, we used to deliver packets going to the host via ipv4_local_delivery

Don't we pass everything up the stack with per endpoint routes? The ipv{4,6}_local_delivery() was used for !per-endpoint routes case to deliver into Pods.

Reply:
Well, I guess that's part of the weird thing here. If you look at the code as it is in master today, you can see that the call to `ipv4_local_delivery` (3 lines below) was outside the `ENABLE_ROUTING` guard. So we were (and still are) calling `ipv4_local_delivery` regardless of the endpoint routes setting it seems :confused:

Reply:
In master this was actually protected by the `is_defined(ENABLE_ROUTING) || hairpin`, so in case of endpoint routes this piece of code is skipped

Reply:
Oh, right :facepalm: I was only looking at the ifdefs.

I think I get the change now. The old check on `ENABLE_ROUTING` (line 975) was useless then, right?

Reply:
The old one was to optimize code for the hairpin case, I guess